File: data-structures/queues/python/queue_priority_OOP.py
# Raspberry Pi Foundation
# Developed to be used alongside Isaac Computer Science,
# part of the National Centre for Computing Education
# Usage licensed under CC BY-SA 4


class Node():
    """A node in a linked list"""

    # Constructor method
    def __init__(self, item_data, item_priority):
        self.data = item_data
        self.priority = item_priority
        self.next = None

    def get_data(self):
        return self.data

    def get_priority(self):
        return self.priority

    def get_next(self):
        return self.next

    def set_next(self, new_next):
        self.next = new_next
        
    def output_node(self):
        """Testing"""
        next_data = None
        if (self.next != None):
            next_data = self.next.data
        print(f"Data: {self.data}, priority: {self.priority}, next: {next_data}")    


class PriorityQueue():
    """A priority queue stored as a linked list"""
    
    # Constructor method
    def __init__(self):
        self.front = None
        self.rear = None

    def get_front(self):
        return self.front

    def set_front(self, new_front):
        self.front = new_front

    def get_rear(self):
        return self.rear

    def set_rear(self, new_rear):
        self.rear = new_rear
    
    def is_empty(self):
        """Check if the queue is empty"""
        if self.front == None:
            return True
        else:
            return False

    def enqueue(self, data, priority):
        """Enqueue an item based on the priority"""
    
        new_node = Node(data, priority)

        # Check if the queue is empty
        if self.is_empty():
            self.front = new_node
            self.rear = new_node
        else:
            # Check if new node should be the new front of queue
            if new_node.get_priority() > self.front.get_priority():
                # Change the current head pointer
                new_node.set_next(self.front)
                self.front = new_node
            # Check if new node should be added to the end of the queue
            elif new_node.get_priority() <= self.rear.get_priority():
                self.rear.set_next(new_node)
                self.rear = new_node
            else:
                # Find the insertion point
                current = self.front
                while(current.get_priority() >= new_node.get_priority()):
                    previous = current
                    current = current.get_next()
                new_node.set_next(current)
                previous.set_next(new_node)
                

    def dequeue(self):
        """Dequeue an item"""
        
        dequeued_item = None
        
        if self.is_empty():
            print("\nQueue is empty - nothing to dequeue")
        else:
             # Get data and change pointer to next item
            dequeued_item = self.front.get_data()
            self.front = self.front.get_next()
            
        return dequeued_item


    def output_queue(self):
        """Output the state of the queue"""

        # Testing
        if self.front != None:
            print("------ State of the queue ------")
            print("Head ", end="")
            current_node = self.front
            current_node.output_node()
            while current_node.get_next() != None:
                current_node = current_node.get_next()
                current_node.output_node()
            print()


def test_enqueue(my_queue):
    print("Adding A with priority 3")
    my_queue.enqueue("A", 3)
    my_queue.output_queue()
    
    print("Adding B with priority 2 - should be behind A")
    my_queue.enqueue("B", 2) 
    my_queue.output_queue()
    
    print("Adding C with priority 3 - should be in front of B")
    my_queue.enqueue("C", 3)
    my_queue.output_queue()
    
    print("Adding D with priority 3 - should be after C")
    my_queue.enqueue("D", 3)
    my_queue.output_queue()
    
    print("Adding E with priority 3 - should be after D")
    my_queue.enqueue("E", 3)
    my_queue.output_queue()
    
    print("Adding F with priority 1 - should be after B")
    my_queue.enqueue("F", 1) 
    my_queue.output_queue()
    
    print("Adding G with priority 4 - should be head of queue")
    my_queue.enqueue("G", 4)
    my_queue.output_queue()
    
    print("Adding H with priority 5 - should be head of queue")
    my_queue.enqueue("H", 5)
    my_queue.output_queue()
    
    print("Adding I with priority 5 - should be after H")
    my_queue.enqueue("I", 5)
    my_queue.output_queue()


def test_dequeue(my_queue):
    print("\n------ Dequeuing 10 items ------\n")
    for i in range(10):
        my_queue.dequeue()
        my_queue.output_queue()

        
def main():
    """Create a priority queue and enqueue and dequeue items"""
    
    print("### Priority queue (OOP) ###\n")

    # Initialise the queue
    my_queue = PriorityQueue()

    # Insert test data into the queue
    test_enqueue(my_queue)
    
    # Remove test data from the queue
    test_dequeue(my_queue)

 
# This code will run if this file is executed directly
# (i.e. not called by another program)
if __name__ == '__main__':
    main()

File: data-structures/queues/c-sharp/queue_priority_OOP.cs
/*
Raspberry Pi Foundation
Developed to be used alongside Isaac Computer Science, part of the National Centre for Computing Education
Usage licensed under CC BY-SA 4

Note: This file is designed to be copied out and compiled on your machine.
In order for it to compile properly you need to ensure that the project name is the same as the "namespace" in this file.
To run this file you need to:
1. Copy the contents
2. Paste them into the C# IDE of your choice (Visual Studio, for example)
3. Change the namespace to match your project (if necessary)
4. Compile the program
5. Run the program
*/

using System;

namespace IsaacCodeSamples
{
    // A class for a node in a linked list
class Node
{
    private string data;
    private int priority;
    private Node next;

    // Constructor method
    public Node(string itemData, int itemPriority)
    {
        data = itemData;
        priority = itemPriority;
    }

    public string GetData()
    {
        return data;
    }

    public int GetPriority()
    {
        return priority;
    }

    public Node GetNext()
    {
        return next;
    }

    public void SetNext(Node newNext)
    {
        next = newNext;
    }

        // Testing
        public void OutputNode()
        {
            string nextData = "null";
            if (next != null)
                nextData = next.data;
            Console.WriteLine($"Data: {data}, priority: {priority}, next: {nextData}");
        }
    }


    // A priority queue stored as a linked list
    class PriorityQueue
    {
        private Node front;
        private Node rear;

        public Node GetFront()
        {
            return front;
        }

        public void SetFront(Node newFront)
        {
            front = newFront;
        }

        public Node GetRear()
        {
            return rear;
        }

        public void SetRear(Node newRear)
        {
            rear = newRear;
        }

        // Check if the queue is empty
        public bool IsEmpty()
        {
            if (front == null)
                return true;
            else
                return false;
        }

        // Enqueue an item based on the priority
        public void Enqueue(string data, int itemPriority)
        {
            Node newNode = new Node(data, itemPriority);

            // Check if the queue is empty
            if (IsEmpty()) {
                front = newNode;
                rear = newNode;
            }
            else {
                // Check if new node should be the new front of queue
                if (newNode.GetPriority() > front.GetPriority()) {
                    // Change the current head pointer
                    newNode.SetNext(front);
                    front = newNode;
                }
                // Check if new node should be added to the end of the queue
                else if (newNode.GetPriority() <= rear.GetPriority()) {
                    rear.SetNext(newNode);
                    rear = newNode;
                }
                else {
                    // Find the insertion point
                    Node current = front;
                    Node previous = null;
                    while (current.GetPriority() >= newNode.GetPriority()) {
                        previous = current;
                        current = current.GetNext();
                    }
                    newNode.SetNext(current);
                    previous.SetNext(newNode);
                }
            }
        }

        // Dequeue an item
        public string Dequeue()
        {
            string dequeuedItem;

            if (IsEmpty()) {
                Console.WriteLine("\nQueue is empty - nothing to dequeue");
                dequeuedItem = "";
            }
            else {
                // Get data and change pointer to next item
                dequeuedItem = front.GetData();
                front = front.GetNext();
            }

            return dequeuedItem;
        }

        // Output the state of the queue
        public void OutputQueue()
        {
            // Testing
            if (front != null) {
                Console.WriteLine("------ State of the queue ------");
                Console.Write("Head ");
                Node currentNode = front;
                currentNode.OutputNode();
                while (currentNode.GetNext() != null) {
                    currentNode = currentNode.GetNext();
                    currentNode.OutputNode();
                }
                Console.WriteLine();
            }
        }
    }


    class Queues
    {
        // The Main method is the entry point for all C# programs
        public static void Main()
        {
            Console.WriteLine("### Priority queue (OOP) ###\n");

            // Initialise the queue
            PriorityQueue myQueue = new PriorityQueue();

            // Insert test data into the queue
            TestEnqueue(myQueue);
            
            // Remove test data from the queue
            TestDequeue(myQueue);
        }

        public static void TestEnqueue(PriorityQueue myQueue)
        {
            Console.WriteLine("Adding A with priority 3");
            myQueue.Enqueue("A", 3);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding B with priority 2 - should be behind A");
            myQueue.Enqueue("B", 2);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding C with priority 3 - should be in front of B");
            myQueue.Enqueue("C", 3);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding D with priority 3 - should be after C");
            myQueue.Enqueue("D", 3);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding E with priority 3 - should be after D");
            myQueue.Enqueue("E", 3);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding F with priority 1 - should be after B");
            myQueue.Enqueue("F", 1);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding G with priority 4 - should be head of queue");
            myQueue.Enqueue("G", 4);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding H with priority 5 - should be head of queue");
            myQueue.Enqueue("H", 5);
            myQueue.OutputQueue();
            
            Console.WriteLine("Adding I with priority 5 - should be after H");
            myQueue.Enqueue("I", 5);
            myQueue.OutputQueue();
        }

        public static void TestDequeue(PriorityQueue myQueue)
        {
            Console.WriteLine("\n------ Dequeuing 10 items ------\n");

            for (int i = 0; i < 10; i++)
            {
                myQueue.Dequeue();
                myQueue.OutputQueue();
            }
        }

    }
}